Python/optimizer_analysis.c

```diff
@@ -533,6 +533,128 @@ remove_unneeded_uops(_PyUOpInstruction *buffer, int buffer_size)
     Py_UNREACHABLE();
 }
 
+/* _PUSH_FRAME/_POP_FRAME's operand can be 0, a PyFunctionObject *, or a
+ * PyCodeObject *. Retrieve the code object if possible.
+ */
+static PyCodeObject *
+get_co(_PyUOpInstruction *op) {
+    assert(op->opcode == _PUSH_FRAME || op->opcode == _POP_FRAME);
+    PyCodeObject *co = NULL;
+    uint64_t operand = op->operand;
+    if (operand == 0) {
+        return NULL;
+    }
+    if (operand & 1) {
+        co = (PyCodeObject *)(operand & ~1);
+    }
+    else {
+        PyFunctionObject *func = (PyFunctionObject *)operand;
+        assert(PyFunction_Check(func));
+        co = (PyCodeObject *)func->func_code;
+    }
+    assert(PyCode_Check(co));
+    return co;
+}
+
+// _PUSH_FRAME is 3 ops in front of _CHECK_STACK_SPACE
+#define _CHECK_STACK_TO_PUSH_FRAME_OFFSET 3
+
+/* Compute the maximum stack space needed for this trace and consolidate
+ * all possible _CHECK_STACK_SPACE ops to one _CHECK_STACK_SPACE_OPERAND.
+ * This calculation is essentially a traversal over the tree of function
+ * calls in this trace.
+ */
+static void
+combine_stack_space_checks(
```

Comment:
Rather than make this a separate pass, could you integrate into `optimize_uops` by adding it to `optimizer_bytecodes.c`.

Reply:
Happy to do this, but to confirm - do you definitely prefer integration into `optimize_uops` over `peephole_opt` or `remove_unneeded_uops`? Any of these would prevent an extra pass.

Reply:
@Fidget-Spinner may also have an opinion here